{% extends '_base.html' %}

{% block content %}

<div class="container">
<h3 align="center">Cluster Across Datasets  &  Generate Mass Spec Database</h3>
</div>
<br>

<div class="container">

	<form method="POST" action="{{ url_for( 'frontend.msdb' ) }}">
    {{ msdb_form.csrf_token }}

  	<div style="margin-left: 20px; margin-right: 20px;" class="panel panel-default">
      <div class="panel-heading" align="center">
        <b>Select Analysis Options</b>
      </div>
      <div class='panel-body'>
        <table class='table table-striped' width=100%> 
          <thead>
              <tr>
                  <th>Analysis Name</th>
                  <th>Analysis Description</th>
              </tr>
          </thead>
          <tbody>
              <tr>
                <td>{{ msdb_form.name(class = 'form-control') }}</td>
                <td>{{ msdb_form.description(class = 'form-control') }}</td>
              </tr>
          </tbody>
        </table>

        <br>
        <br>

        <div class="panel panel-default">
            <div class="panel-heading" align="center">
                <b>Step One: Error Correction by High-Similarity Read Clustering</b>
            </div>
            <div class="panel-body">
                <table class='table table-striped' width=100%>
                    <thead>
                        <tr>
                            <th width=10%>Cluster On</th>
                            <th width=15%>Cluster Homology</th>
                            <th width=18%>Cluster Algorithm</th>
                            <th width=15%>Minimum # Reads</th>
                            <th width=20%>Max Seqs To Report Per Cluster</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                          <td>{{ msdb_form.error_correct_cluster_on() }}</td>
                          <td>{{ msdb_form.error_correct_cluster_percent(class="form-control", type="number", step = "0.01") }}</td>
                          <td>{{ msdb_form.error_correct_cluster_algorithm(class="form-control") }}</td>
                          <td>{{ msdb_form.error_correct_read_cutoff(class = 'form-control') }}</td>
                          <td>{{ msdb_form.error_correct_max_sequences_per_cluster_to_report(class = 'form-control')}}</td>
                        </tr>
                    </tbody>
                </table>
                <table>
                    <tr>
                        <td>Require these regions be annotated: {{ msdb_form.require_annotations() }}</td>
                    </tr>
                </table>
            </div>
        </div>


        <div class="panel panel-default">
            <div class="panel-heading" align="center">
                <b>Step Two: Clonotyping by High-Sensitivity Clustering On CDR3 AA or NT</b>
            </div>
            <div class="panel-body">
                <table class='table table-striped' width=100%>
                    <thead>
                        <tr>
                            <th width=10%>Cluster On</th>
                            <th width=15%>Cluster Homology</th>
                            <th width=18%>Cluster Algorithm</th>
                            <th width=15%>Minimum # Reads</th>
                            <th width=20%>Max Seqs To Report Per Cluster</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                          <td>{{ msdb_form.clonotyping_cluster_on() }}</td>
                          <td>{{ msdb_form.clonotyping_cluster_percent(class="form-control", type="number", step = "0.01") }}</td>
                          <td>{{ msdb_form.clonotyping_cluster_algorithm(id='cluster_algorithm_select', class='form-control', onchange="hide_linkage(this)") }}<br><br><div id="linkage_select">Linkage: {{msdb_form.clonotyping_cluster_linkage()}}<br><br>Warning: HUGE MEMORY USAGE. Only use with small files (>10K annotations) or it's likely to fail.</div><br><a href="http://drive5.com/usearch/manual/uclust_algo.html">Greedy Info</a><br><a href="http://drive5.com/usearch/manual/agg.html">Agglomerative Info</a><br><br></td>
                          <td>{{ msdb_form.clonotyping_read_cutoff(class = 'form-control') }}</td>
                          <td>{{ msdb_form.clonotyping_max_sequences_per_cluster_to_report(class = 'form-control')}}</td>
                        </tr>
                    </tbody>
                </table>            </div>
        </div>

          <br>

        <div class="panel panel-default">
            <div class="panel-heading" align="center">
                <b>Step Three: Post-Analysis Routines To Optimize Mass Spec Search Database</b>
            </div>
            <div class="panel-body">
                <table class='table table-striped' width=100%>
                  <thead>
                      <tr>
                          <th width=12%>Generate Fasta File</th>
                          <th width=20%>Append C-terminal Peptides</th>
                          <th width=16%>Rescue N-terminal Peptides</th>
                          <th width=16%>Rescue C-terminal Peptides</th>
                          <th width=15%>Confirm Isotype Calls</th>
                      </tr>
                  </thead>
                  <tbody>
                      <tr>
                          <td>{{ msdb_form.generate_fasta_file(class = '') }}</td>
                          <td>{{ msdb_form.append_cterm_peptides(class = 'append_cterm_peptides') }}
                          <td>{{ msdb_form.rescue_n_terminal_peptides(class = '', disabled=True) }}</td>
                          <td>{{ msdb_form.rescue_c_terminal_peptides(class = '', disabled=True) }}</td>
                          <td>{{ msdb_form.confirm_isotype_calls(class = '', disabled=True) }}</td>
                            </td>
                      </tr>
                  </tbody>
                </table>
                <div id="collapse" class="container collapse append_cterm_peptides in" style="display: none; width: 80%;" align="center">
                    <br>
                        <b>This preliminary 'Append Cterm Peptides' function artificially adds missing peptides to the end of the AA sequences:</b>
                        <br><br>
                        <p>To the end of IgL Full AA Seqs, 'GQPK' will be added.</p>
                        <p>To the end of IgK Full AA Seqs, 'R'.</p>
                        <p>To the end of IgH Full AA Seqs, 'ASTK'.</p>
                        <p>#Eventually this will be rewritten to recover exact translation from 3' of raw reads.</p>
                    <br>
                </div>
            </div>
        </div>


      </div>
    </div>

    <br>
    <br>

    <div style="margin-left: 20px; margin-right: 20px;" class="panel panel-default">
      <div class="panel-heading" align="center">
        <b>Select Compatible Annotation Files</b>
      </div>
      <div class='panel-body'>
        <p align="center">Caution: Only Group One Type Of File, Paired Or Single-Chain, In The Same Run At This Time</p>
        <p align="center">If Paired Files Are Run, Heavy Chain Annotations Will Be Used For Cluster On + Require Annotated Fields</p>
        <br>
        <table id="datasets" class="display table-striped table table-bordered" cellspacing="0" width="100%">
          <thead>
            <tr>
              <th class='no-sort' style="width:20px;""> </th>
              <th>Dataset #</th>
              <th>Name</th>
              <th>Description</th>
              <th>Projects</th>
            </tr>
          </thead>
          <tbody>
            {% for dataset, files in dataset_file_dict.iteritems() %}
              <tr>
                <td {% if files != [] -%} class='details-control' {%- endif -%} ></td>
                <td><a href="{{url_for('.dataset', id=dataset.id)}}">{{dataset.id}}</a></td>
                <td><a href="{{url_for('.dataset', id=dataset.id)}}">{{dataset.name}}</a></td>
                <td>{{dataset.description}}</td>
                <td>{{dataset.projects|map(attribute='name')|join(',')}}</td>
              </tr>
            {% endfor %}
            
          </tbody>
        </table>
      </div>
    </div>

    <br>

    <div style="margin-left: 20px; margin-right: 20px;" class="panel panel-default">
      <div class="panel-heading" align="center">
        <input type='submit' class="btn btn-default form-control" value="Generate Clustered Annotation / MSDB Fasta Files"></td>
      </div>
    </div>



  </form>
</div>



<br>
<br>
<br>
<br>
<br>
<br>


<script>

$('input[type=checkbox]').on('change', function () {
    
    if (!this.checked){
      $('.collapse.' + $(this).attr('class')).slideUp();
    }else{
      $('.collapse.' + $(this).attr('class')).slideDown();
    }
});


var table = $('#datasets').DataTable({
  columnDefs: [
    { targets: 'no-sort', orderable: false }
    ],
    order: [[ 1, 'desc' ]], 
    pageLength:10
});


var table_data = [
{%- for dataset, files in dataset_file_dict.iteritems() -%}
    `
    <!-- HTML to write -->
    <table class="table table-striped">
      <thead>
        <tr>
          <th></th>
          <th>File ID</th>
          <th>Name</th>
          <th>File Type</th>
          <th>Available</th>
          <th>File Size</th>
          <th>Lines</th>
          <th>Created</th>
        </tr>
      </thead>
      <tbody>
        {%- for file in files|sort(attribute='id', reverse=True) -%}
          <tr>
            <td><input name="file_ids" type="checkbox" value="{{file.id}}"</td>
            <td><a href="{{url_for('frontend.file', id=file.id)}}" >{{file.id}}</a> </td>
            <td>{{file.name}}</td>
            <td>{{file.file_type}}</td>
            {%- if file.available -%}
              <td>YES</td>
            {%- else -%}
              <td>NO</td>
            {%- endif -%}
            {%- if file.file_size -%}
              <td>{{file.file_size|filesizeformat}}</td>
            {%- elif file.available -%}
              <td>0 bytes</td>
            {%- else -%}
              <td></td>
            {%- endif -%}
            {%- if file.line_count -%}
              <td>{{file.line_count_string}}</td>
            {%- elif file.available -%}
              <td>0</td>
            {%- else -%}
              <td></td>
            {%- endif -%}
            <td>{{file.created}}</td>
          </tr>
        {%- endfor -%}
      </tbody>
    </table>`,
{%- endfor -%}
  ''
]
table.rows().every( function () {
    this.child( table_data[this.index()] );
} );

$('#datasets tbody').on( 'click', 'td.details-control', function () {
    var child = table.row( this ).child;
  var tr = $(this).closest('tr');
    
    if ( child.isShown() ) {
        child.hide();
        tr.removeClass('shown');
    }
    else {
        child.show();
        tr.addClass('shown')
    }
} );




function hide_linkage() {
    var selectBox = document.getElementById("cluster_algorithm_select");
    var selected = selectBox.options[selectBox.selectedIndex].value;

    if(selected === 'agglomerative'){
        $('#linkage_select').show();
    }
    else{
        $('#linkage_select').hide();
    }
}

$(document).ready( function() {
  hide_linkage();
}); 


</script>

{%endblock content %}































